src/mongo_odm/deserializing_cursor.hpp

```diff
@@ -0,0 +1,98 @@
+// Copyright 2016 MongoDB Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <mongo_odm/config/prelude.hpp>
+
+#include <iostream>
+
+#include <bsoncxx/builder/basic/document.hpp>
+#include <bsoncxx/stdx/optional.hpp>
+#include <mongocxx/cursor.hpp>
+
+#include <bson_mapper/mapping_functions.hpp>
+
+namespace mongo_odm {
+MONGO_ODM_INLINE_NAMESPACE_BEGIN
+
+/**
+ * A class that wraps a mongocxx::cursor. It provides an iterator that deserializes the
+ * documents
+ * yielded by the underlying mongocxx cursor.
+ *
+ * TODO what to do if some documents fail deserialization?
+ */
+template <class T>
+class deserializing_cursor {
+   public:
+    deserializing_cursor(mongocxx::cursor&& c) : _c(std::move(c)) {
+    }
+
+    class iterator;
+
+    iterator begin() {
+        return iterator(_c.begin());
+    }
+
+    iterator end() {
+        return iterator(_c.end());
+    }
+
+   private:
+    mongocxx::cursor _c;
+};
+
+template <class T>
+class deserializing_cursor<T>::iterator : public std::iterator<std::input_iterator_tag, T> {
```

Comment:
I don't know how much overhead this would incur, but this is creating a new BSONInputArchive for every BSON object we get back from the cursor. 

The BSONInputArchive supports unbounded streams so this deserializing cursor can hold a single BSONInputArchive and the `*` overload can be a matter of writing the BSON data to the stream associated with the archive and calling the archive's `()` overload. You definitely don't have to do this now, but it might be worth putting in a TODO for when I start doing perf work later on.

A similar kind of optimization could probably be done with the serializing_iterator.